cpp/src/arrow/compute/exec/window_functions/bit_vector_navigator.h

```diff
@@ -0,0 +1,513 @@
+                 block_index < static_cast<int64_t>(container_->mid_counts_.size()));
+    constexpr int64_t words_per_block =
+        BitVectorWithCountsBase::kBitsPerBlock / BitVectorWithCountsBase::kBitsPerWord;
+    int64_t word_begin = block_index * words_per_block;
+    int64_t word_end = std::min(word_count(), word_begin + words_per_block);
+
+    const uint64_t* words = container_->bits_.data();
+    uint16_t* counters = container_->mid_counts_.data();
+
+    uint16_t count = 0;
+    for (int64_t word_index = word_begin; word_index < word_end; ++word_index) {
+      counters[apply_stride_and_offset(word_index)] = count;
+      count += static_cast<uint16_t>(
+          ARROW_POPCOUNT64(words[apply_stride_and_offset(word_index)]));
+    }
+  }
+
+  void BuildTopCounts(int64_t block_index_begin, int64_t block_index_end,
+                      int64_t initial_count = 0) {
+    const uint64_t* words = container_->bits_.data();
+    int64_t* counters = container_->top_counts_.data();
+    const uint16_t* mid_counters = container_->mid_counts_.data();
+
+    int64_t count = initial_count;
+
+    for (int64_t block_index = block_index_begin; block_index < block_index_end - 1;
+         ++block_index) {
+      counters[apply_stride_and_offset(block_index)] = count;
+
+      constexpr int64_t words_per_block =
+          BitVectorWithCountsBase::kBitsPerBlock / BitVectorWithCountsBase::kBitsPerWord;
+
+      int64_t word_begin = block_index * words_per_block;
+      int64_t word_end = std::min(word_count(), word_begin + words_per_block);
+
+      count += mid_counters[apply_stride_and_offset(word_end - 1)];
+      count += ARROW_POPCOUNT64(words[apply_stride_and_offset(word_end - 1)]);
+    }
+    counters[apply_stride_and_offset(block_index_end - 1)] = count;
+  }
+
+  // Position of the nth bit set (input argument zero corresponds to the first
+  // bit set).
+  //
+  int64_t Select(int64_t rank) const {
+    if (rank < 0) {
+      return BeforeFirstBit();
+    }
+    if (rank >= pop_count()) {
+      return AfterLastBit();
+    }
+
+    constexpr int64_t bits_per_block = BitVectorWithCountsBase::kBitsPerBlock;
+    constexpr int64_t bits_per_word = BitVectorWithCountsBase::kBitsPerWord;
+    constexpr int64_t words_per_block = bits_per_block / bits_per_word;
+    const int64_t* top_counters = container_->top_counts_.data();
+    const uint16_t* mid_counters = container_->mid_counts_.data();
+    const uint64_t* words = container_->bits_.data();
+
+    // Binary search in top level counters.
+    //
+    // Equivalent of std::upper_bound() - 1, but not using iterators.
+    //
+    int64_t begin = 0;
+    int64_t end = block_count();
+    while (end - begin > 1) {
+      int64_t middle = (begin + end) / 2;
+      int reject_left_half =
+          (rank >= top_counters[apply_stride_and_offset(middle)]) ? 1 : 0;
+      begin = begin + (middle - begin) * reject_left_half;
+      end = middle + (end - middle) * reject_left_half;
+    }
```

Comment:
so unlike `std::upper_bound` this is a branchless binary search which will be good when the window is small and bad when the window is large. Do we know it always pays off?